### DISCUSSION TITLE
Vehicles don't follow set path

### DISCUSSION THREAD

**Discussion by 2Innovative** [carla-simulator/carla#5263] [Q&A]
(CARLA 0.9.13 release, Win10)
So, I did this tutorial:
https://carla.readthedocs.io/en/docs-preview/tuto_G_traffic_manager/
But the cars barely stick to the path. On some paths they just take random turns and on other paths they even crash into trees on the sidewalk.
I can't find the problem, the tutorial seems pretty straight forward, just tell a car to go to a waypoint and CARLA does the rest but it doesn't work like that for me.
![crash01](https://user-images.githubusercontent.com/93536264/158571417-b04c0f1e-4f62-4a28-a182-e423b0ce9db6.PNG)

This is the code, of course I already played around with it but it should be similar enough to the tutorial.

import glob
import os
import sys

try:
    sys.path.append(glob.glob('../carla/dist/carla-*%d.%d-%s.egg' % (
        sys.version_info.major,
        sys.version_info.minor,
        'win-amd64' if os.name == 'nt' else 'linux-x86_64'))[0])
except IndexError:
    pass
    
import carla
import random

# Connect to the client and retrieve the world object
client = carla.Client('localhost', 2000)
world = client.get_world()

# Set up the simulator in synchronous mode
settings = world.get_settings()
settings.synchronous_mode = True # Enables synchronous mode
settings.fixed_delta_seconds = 0.05
world.apply_settings(settings)

# Set up the TM in synchronous mode
traffic_manager = client.get_trafficmanager()
traffic_manager.set_synchronous_mode(True)

# We will aslo set up the spectator so we can see what we do
spectator = world.get_spectator()

spawn_points = world.get_map().get_spawn_points()

# Draw the spawn point locations as numbers in the map
for i, spawn_point in enumerate(spawn_points):
    world.debug.draw_string(spawn_point.location, str(i), life_time=300)

# Select some models from the blueprint library
models = ['dodge', 'audi', 'model3', 'mini', 'mustang', 'lincoln', 'prius', 'nissan', 'crown', 'impala']
blueprints = []
for vehicle in world.get_blueprint_library().filter('*vehicle*'):
    if any(model in vehicle.id for model in models):
        blueprints.append(vehicle)

# Route 1
spawn_point_1 =  spawn_points[32]
# Create route 1 from the chosen spawn points
route_1_indices = [129, 28, 124, 33, 97, 119, 58, 154, 147]
route_1 = []
for ind in route_1_indices:
    route_1.append(spawn_points[ind].location)

# Route 2
spawn_point_2 =  spawn_points[149]
# Create route 2 from the chosen spawn points
#route_2_indices = [21, 76, 38, 34, 90, 3]
#I played around with different paths
route_2_indices = [21, 105, 52, 104, 141, 109, 13, 80, 0, 104, 140, 10, 149]
route_2 = []
for ind in route_2_indices:
    route_2.append(spawn_points[ind].location)

# Now let's print them in the map so we can see our routes
world.debug.draw_string(spawn_point_1.location, 'Spawn point 1', life_time=300, color=carla.Color(255,0,0))
world.debug.draw_string(spawn_point_2.location, 'Spawn point 2', life_time=300, color=carla.Color(0,0,255))

for ind in route_1_indices:
    spawn_points[ind].location
    world.debug.draw_string(spawn_points[ind].location, str(ind), life_time=300, color=carla.Color(255,0,0))

for ind in route_2_indices:
    spawn_points[ind].location
    world.debug.draw_string(spawn_points[ind].location, str(ind), life_time=300, color=carla.Color(0,0,255))

# Set delay to create gap between spawn times
spawn_delay = 20
counter = spawn_delay

# Set max vehicles (set smaller for low hardward spec)
max_vehicles = 1 #having several cars with auto lane change makes cars go different routes 
#but not even a single car sticks to the given path
# Alternate between spawn points
#for test purposes I don't want them to alternate for now   
alt = False

# In synchronous mode, we need to run the simulation to fly the spectator
while True:
    world.tick()
    
    n_vehicles = len(world.get_actors().filter('*vehicle*'))
    vehicle_bp = random.choice(blueprints)

    # Spawn vehicle only after delay
    if counter == spawn_delay and n_vehicles < max_vehicles:
        # Alternate spawn points
        if alt:
            vehicle = world.try_spawn_actor(vehicle_bp, spawn_point_1)
        else:
            vehicle = world.try_spawn_actor(vehicle_bp, spawn_point_2)

        if vehicle: # IF vehicle is succesfully spawned
            vehicle.set_autopilot(True) # Give TM control over vehicle

            # Set parameters of TM vehicle control, we don't want lane changes
            traffic_manager.update_vehicle_lights(vehicle, True)
            traffic_manager.random_left_lanechange_percentage(vehicle, 0)
            traffic_manager.random_right_lanechange_percentage(vehicle, 0)
            #allowing auto lane change solves some problems but creates others
            traffic_manager.auto_lane_change(vehicle, False)

            # Alternate between routes
            if alt:
                traffic_manager.set_path(vehicle, route_1)
                #alt = False
            else:
                traffic_manager.set_path(vehicle, route_2)
                #alt = True

            vehicle = None

        counter -= 1
    elif counter > 0:
        counter -= 1
    elif counter == 0:
        counter = spawn_delay

**Comment by 2Innovative** [comment]
So, if the release version is behaving like this, you should try the editor version. 
Yes, there is the build process but my experience was that the editor behaves less defiant.

**Reply by agtbaskara** [reply]
Can you share a link or guide for the editor version of traffic manager? I also found that the set_path is unreliable and currently trying to find the solution for this.

**Reply by 2Innovative** [reply]
> Can you share a link or guide for the editor version of traffic manager? I also found that the set_path is unreliable and currently trying to find the solution for this.

Here you go:
https://carla.readthedocs.io/en/0.9.13/build_windows/

**Reply by agtbaskara** [reply]
Is the 'editor' version mean the latest Carla version from the master branch?

**Reply by 2Innovative** [reply]
The term 'version' might be a little bit misleading.
You can Build CARLA yourself from the source code by following the steps in the link. By doing so, you include the Unreal Engine which comes with an editor.
So, if you run the CARLA (version) that you have build yourself, it will start an editor where you can make changes and also (actually) run the simulation. 
If you just download the 'release' version of CARLA (from the Quick start guide) there is no editor included but you can start and run CARLA right away.
I hope that helps.

**Comment by NunWang** [comment]
yes,  I use carla 0.9.15, and I find that carla.TrafficManager.set_path is still no useful.
